broadcaster/_backends/pulsar.py

```diff
@@ -1,77 +1,119 @@
+            logger.info(f"Subscribed to channel: {channel}")
+
 
     async def unsubscribe(self, channel: str) -> None:
-        # Similarly, we don't need to do anything here
-        pass
+        if channel in self._consumers:
+            consumer = self._consumers.pop(channel)
+            try:
+                await anyio.to_thread.run_sync(consumer.close)
+            except ValueError:
+                logger.warning(f"Consumer for channel {channel} was not in the client's list")
+            except Exception as e:
+                logger.error(f"Error closing consumer for channel {channel}: {e}")
+            logger.info(f"Unsubscribed from channel: {channel}")
+
 
     async def publish(self, channel: str, message: typing.Any) -> None:
-        encoded_message = f"{channel}:{message}".encode("utf-8")
-        await anyio.to_thread.run_sync(lambda: self._producer.send(encoded_message))
+        if channel not in self._producers:
+            self._producers[channel] = await anyio.to_thread.run_sync(
+                lambda: self._client.create_producer(channel)
+            )
+        encoded_message = str(message).encode("utf-8")
+        await anyio.to_thread.run_sync(lambda: self._producers[channel].send(encoded_message))
+        logger.info(f"Published message to channel {channel}: {message}")
 
     async def next_published(self) -> Event:
-        while True:
-            try:
-                msg = await anyio.to_thread.run_sync(self._consumer.receive)
-                channel, content = msg.data().decode("utf-8").split(":", 1)
-                await anyio.to_thread.run_sync(lambda: self._consumer.acknowledge(msg))
-                return Event(channel=channel, message=content)
-
-            except anyio.get_cancelled_exc_class():
-                # cancellation
-                logging.info("next_published task is being cancelled")
-                raise
+        return await self._shared_queue.get()
 
-            except Exception as e:
-                logging.error(f"Error in next_published: {e}")
-                raise
+    async def _receiver(self, channel: str, consumer: pulsar.Consumer) -> None:
+        try:
+            while True:
+                try:
+                    msg = await anyio.to_thread.run_sync(consumer.receive)
+                    content = msg.data().decode("utf-8")
+                    await anyio.to_thread.run_sync(consumer.acknowledge, msg)
+                    await self._shared_queue.put(Event(channel=channel, message=content))
+                    logger.info(f"Received message from channel {channel}: {content}")
+                except Exception as e:
+                    logger.error(f"Error receiving message from channel {channel}: {e}")
+        except asyncio.CancelledError:
+            logger.info(f"Receiver for channel {channel} was cancelled")
+        finally:
+            await anyio.to_thread.run_sync(consumer.close)
```

Comment:
Closing the consumer is done also in the disconnect. That will cause consumers to be closed twice, is it fine from the Pulsar's consumer interface?

Reply:
it's actually fine for consumers to be closed both in the receiver task's cleanup and in the disconnect() method.